### ISSUE TITLE
Can't specify will delay on CONNECT packet with auth_on_publish_m5 hook enabled

### ISSUE THREAD

**Issue by aj9411** [vernemq/vernemq#1998]
### Environment

- VerneMQ Version: 1.12.5
- OS: Alpine Linux v3.15
- Erlang/OTP version (if building from source): 23.3.4.0-alpine
- Cluster size/standalone: standalone

### Expected behaviour

Not much here, when I send a CONNECT packet with a will message that includes a delay, I expect it to just work as outlined on the MQTT v5 spec.  Sadly, it is not working for me now.

### Actual behaviour

When I send a CONNECT packet with a will message that includes a delay, I receive a CONNACK packet with a reason code of 'not_authorized'. I traced the call to the auth_on_publish_m5 hook using vmq-admin, see output below:

2022-06-07T18:17:22Z <8526.2958.3> Calling auth_on_publish_m5(undefined,{[],
<<"{CLIENT-ID_REDACTED}">>},0,test/sampleTopicef3d88ba-5c6b-41f0-bb2c-11547b8ebefe,false) with payload:
   testPayload
   with properties: #{p_message_expiry_interval => 3600,
                                p_will_delay_interval => 2}
2022-06-07T18:17:29Z <8526.2958.3> Hook returned {error,error}
2022-06-07T18:17:29Z <8526.2958.3> MQTT SEND: CID: "CLIENT-ID_REDACTED" CONNACK(sp: 0, rc: not_authorized(135))

I am running VerneMQ on my development machine, and I set it up to call localhost endpoints for its webhooks. The modifiers response from the auth_on_publish_m5 hook has the p_will_delay_interval on it, as I copy the request properties into the modifiers properties. The will delay interval comes in the request to auth_on_publish_m5 . If I null the p_will_delay_interval property on the response, then I get a successful CONNACK packet, but the will delay is not respected. I think I am getting the 'not_authorized' response because P_WILL_DELAY_INTERVAL is not on `vmq_plugin_util.erl` at `modifiers(auth_on_publish_m5)`.  I tried overriding that file with the addition of P_WILL_DELAY_INTERVAL and now I get a successful CONNACK, but also with the will delay interval not being respected, maybe I am missing another needed alteration?

- Please include minimal steps to reproduce the issue, if possible.

1. Spin up vernemq with an auth_on_publish_m5 hook
2. Send a CONNECT packet with a will message and a will delay
3. See the response CONNACK packet with the 'not_authorized' reason code

**Comment by aj9411** [comment]
Sure, it is:

2022-06-10 17:19:27.556 [error] <0.1965.0>@vmq_mqtt5_fsm:auth_on_publish:1173 can't auth publish [undefined,{[],<<"CLIENT_ID_REDACTED">>},0,[<<"test">>,<<"sampleTopicd0115ff5-e96b-4348-8a87-e8fdac0afad1">>],<<"testPayload">>,false,#{p_message_expiry_interval => 604800,p_will_delay_interval => 2}] due to error
2022-06-10 17:25:09.785 [error] <0.2550.0>@vmq_plugin_util:val_properties:187 property not allowed p_will_delay_interval 2

**Comment by ioolkos** [comment]
@aj9411 Thanks, coming back to this to see whether we need to actually fix something.

Note that there should never be a will delay property in a PUBLISH packet (`auth_on_publish` hook must be triggered by a PUBLISH packet not a CONNECT packet). Currently I think you might be doing something unexpected by included those properties in the PUBLISH packet.
Can you confirm that understanding?

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by aj9411** [comment]
Right, a will delay property should never be present on a PUBLISH packet. But I opened this issue regarding a CONNECT packet, I don't have any problems with PUBLISH packets. Shouldn't the `auth_on_publish_m5` hook be called on CONNECT packets _including a WILL_? By including a will, I am asking as a client to publish something at a later time, correct? So, doesn't it make sense that the `auth_on_publish_m5` hook is called in this case so that the topic can be modified and what not?

Can I provide some other logs or evidence to prove that I am not doing something unexpected myself?

**Comment by ioolkos** [comment]
I think an exact client command would help me (with mosquitto_sub or a similar client).

Note that `auth_on_publish` will be called on that last will message, but at session termination, not at connect time.

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by mths1** [comment]
Hello @ioolkos : I have a similar (but different issue):

- auth_on_publish_m5 gets the p_will_delay interval as property. As far as I understand @aj9411 tries to return that as a modifer which is not allowed. Of course, one could argue if that property could be filtered beforhand.

my issue is:

Similar to @aj9411 the auth_on_publish_m5 is called **immediately** and not when the last will message is actually published. To reproduce:

`mosquitto_pub -t test -m 123 --will-topic test2 --will-payload gone --will-qos 1 -V 5 -D will will-delay-interval 20 -u joe -P secret --repeat 100 --repeat-delay 100 -x 5
`
will cause the following 

```
hook: auth_on_register_m5
  data:  {'peer_addr': '127.0.0.1', 'peer_port': 59564, 'mountpoint': '', 'client_id': 'anon-Pg4URyvMwXxFT7SKJ79egt4egNNd', 'username': 'joe', 'password': 'secret', 'clean_start': True, 'properties': {'p_receive_max': 20, 'p_session_expiry_interval': 5}}
127.0.0.1:59014 - - [14/Apr/2023 09:23:20] "HTTP/1.1 POST /myendpoints" - 200 OK
hook: auth_on_publish_m5
  data:  {'username': 'joe', 'mountpoint': '', 'client_id': 'anon-Pg4URyvMwXxFT7SKJ79egt4egNNd', 'qos': 1, 'topic': 'test2', 'payload': 'Z29uZQ==', 'retain': False, 'properties': {'p_will_delay_interval': 20}}
127.0.0.1:59014 - - [14/Apr/2023 09:23:20] "HTTP/1.1 POST /myendpoints" - 200 OK
hook: auth_on_publish_m5
  data:  {'username': 'joe', 'mountpoint': '', 'client_id': 'anon-Pg4URyvMwXxFT7SKJ79egt4egNNd', 'qos': 0, 'topic': 'test', 'payload': 'MTIz', 'retain': False, 'properties': {}}

```

as you can see auth_on_publish_m5 is called immediately after register (topic test2), before the auth of the actual payload happens (test).

**Question:** 
What is the correct behaviour? I can see arguments for both. The will comes with the connect packet. So it feels right to also check the authorization at that point. On the other hand, it might be published years later, so it also feels right to check it (maybe again?) on publish...

**Comment by ioolkos** [comment]
@mths1 @aj9411 my information about the last will publish being checked at session termination was obviously wrong then. I apologise for that!
The authorization context is this: the LW will be published by the broker on a topic that is chosen by the client. Who is allowed what and when then.
Technically, it's simpler to check the last will publish authorization at CONNECT. 
It's probably not mandatory in the spec to authorize a last will topic in the server. But we do it; and then the question is: how to tell the client about this. Instead of disconnecting, we should probably use Connect Response Code 144. ("Topic Name Invalid")
Weirdly enough, the description to that code reads:

> The Will Topic Name is not malformed, but is not accepted by this Server.

Also, we should fix the opening issue, I guess. Modifying the last will delay should be possible in a plugin.

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.
